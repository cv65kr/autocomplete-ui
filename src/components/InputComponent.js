import {h} from "preact";
import {actions} from "../actions";
import { connect } from 'unistore/preact'

/**
 * Input Component
 *
 * @param htmlNodeInheritProps
 * @param searchAction
 * @param keyDownAction
 * @param focusOutAction
 * @constructor
 */
export const InputComponent = ({
   /** component props */
   htmlNodeInheritProps,

   /** actions */
   searchAction,
   keyDownAction,
   focusOutAction
}) => <input
    {...htmlNodeInheritProps}
    autocomplete="off"
    spellCheck="false"
    role="combobox"
    aria-autocomplete="list"
    aria-expanded="false"
    aria-owns="apisearch-listbox"
    data-search="Apisearch-autocomplete"

    onInput={searchAction}
    onKeyDown={keyDownAction}
    onBlur={focusOutAction}
/>;

/**
 * Input connector
 */
const ConnectedInput = connect('', actions)(InputComponent);

export default ConnectedInput;