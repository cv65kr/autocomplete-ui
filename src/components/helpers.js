import {h} from "preact";
import {groupBy, sortBy} from "lodash";
import Hogan from "hogan.js";


/**
 * Compile template
 * @param template
 * @param data
 */
export const renderTemplate = (template, data) => {
    let compiledTemplate = Hogan.compile(template);

    return {
        __html: compiledTemplate.render(data)
    }
};

/**
 * Set active index
 *
 * @param items
 * @param currentCursorIndex
 * @returns {{}}
 */
export const createDatasetByItemTypeAndSetActiveIndex = (
    items,
    currentCursorIndex
) => {
    let sortedItemsByType = sortBy(items, 'uuid.type');
    let itemsWithActiveIndex = sortedItemsByType.map((item, index) =>
        ({
            ...item,
            navigationIndex: (index + 1),
            isActive: ((index + 1) === currentCursorIndex)
        })
    );

    return groupBy(itemsWithActiveIndex, 'uuid.type');
};

/**
 * Build the poweredBy icon
 * It can be a boolean:
 *   if (false) do now show a poweredBy
 *   if (true) assign the default poweredBy
 *   else render a custom poweredBy template string
 *
 * @param poweredBy
 * @returns {XML}
 */
export const buildPoweredBy = (poweredBy = false) => {
    if (false === poweredBy) return null;

    if (true === poweredBy) {
        poweredBy = `<a href="http://apisearch.io">
            <img 
                src="http://apisearch.io/assets/media/powered-by.svg" 
                alt="Powered by Apisearch" 
                width="100px"
            />
        </a>`;
    }

    return <div
        className="as-poweredBy"
        dangerouslySetInnerHTML={
            renderTemplate(poweredBy, null)
        }
    />
};