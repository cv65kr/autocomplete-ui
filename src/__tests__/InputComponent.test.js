import {h} from 'preact';
import {shallow} from 'preact-render-spy'
import {InputComponent} from '../components/InputComponent';

const datasets = [
    {
        type: 'mock',
        template: {
            header: '<h3>Mock Dataset</h3>',
            item: '<p>{{metadata.name}}</p>'
        }
    }
];

describe('<InputComponent />', () => {
    it('should render', () => {
        const tree = shallow(
            <InputComponent htmlNodeInheritProps={{
                id: 'apisearch-input-id',
                class: 'some-class-name'
            }} />
        );

        /**
         * Expect initial store state
         */
        expect(tree).toMatchSnapshot();
    });
});