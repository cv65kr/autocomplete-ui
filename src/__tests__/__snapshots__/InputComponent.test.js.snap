// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputComponent /> should render 1`] = `
preact-render-spy (1 nodes)
-------
<input
  id="apisearch-input-id"
  class="some-class-name"
  autocomplete="off"
  spellCheck="false"
  role="combobox"
  aria-autocomplete="list"
  aria-expanded="false"
  aria-owns="apisearch-listbox"
  data-search="Apisearch-autocomplete"
 />

`;
